In [29]:
import pandas as pd #importing the required libraries
import numpy as np
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

In [30]:
data = pd.read_csv("/content/sales_data_with_discounts.csv") #load the dataset into environment
data.head(11)

Date       Day  SKU City  Volume       BU    Brand        Model  \
0   01-04-2021  Thursday  M01    C      15  Mobiles    RealU        RU-10   
1   01-04-2021  Thursday  M02    C      10  Mobiles    RealU    RU-9 Plus   
2   01-04-2021  Thursday  M03    C       7  Mobiles     YouM        YM-99   
3   01-04-2021  Thursday  M04    C       6  Mobiles     YouM   YM-99 Plus   
4   01-04-2021  Thursday  M05    C       3  Mobiles     YouM       YM-98    
5   01-04-2021  Thursday  M06    C       3  Mobiles    RealU         RU-9   
6   01-04-2021  Thursday  M07    C       3  Mobiles  Sumsang         S-20   
7   01-04-2021  Thursday  M08    C       2  Mobiles  Sumsang         S-21   
8   01-04-2021  Thursday  M09    C       2  Mobiles   Orange         O-10   
9   01-04-2021  Thursday  M10    C       3  Mobiles   Orange         O-11   
10  01-04-2021  Thursday  F01    C      13     FMCG   Babaji  Babaji Oil    

    Avg Price  Total Sales Value  Discount Rate (%)  Discount Amount  \
0       12100             181500          11.654820     21153.498820   
1       10100             101000          11.560498     11676.102961   
2       16100             112700           9.456886     10657.910157   
3       20100             120600           6.935385      8364.074702   
4        8100              24300          17.995663      4372.946230   
5        8100              24300          16.996489      4130.146805   
6       49100             147300           9.228812     13594.039719   
7       54100             108200           5.553719      6009.124321   
8       55100             110200           7.410104      8165.935144   
9       60100             180300           9.928444     17900.983733   
10        400               5200          18.869242       981.200566   

    Net Sales Value  
0     160346.501180  
1      89323.897039  
2     102042.089843  
3     112235.925298  
4      19927.053770  
5      20169.853195  
6     133705.960281  
7     102190.875679  
8     102034.064856  
9     162399.016267  
10      4218.799434

# Descriptive Analytics for Numerical Columns

In [31]:
data.info() #To get the information obout data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450 entries, 0 to 449
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Date               450 non-null    object 
 1   Day                450 non-null    object 
 2   SKU                450 non-null    object 
 3   City               450 non-null    object 
 4   Volume             450 non-null    int64  
 5   BU                 450 non-null    object 
 6   Brand              450 non-null    object 
 7   Model              450 non-null    object 
 8   Avg Price          450 non-null    int64  
 9   Total Sales Value  450 non-null    int64  
 10  Discount Rate (%)  450 non-null    float64
 11  Discount Amount    450 non-null    float64
 12  Net Sales Value    450 non-null    float64
dtypes: float64(3), int64(3), object(7)
memory usage: 45.8+ KB


By considering the information of the dataset, I conclude that Volume, Avg Price, Total Sales Value, Discount Rate (%),	Discount Amount,	Net Sales Value are the numerical Columns .

In [32]:
data.describe() #To get the Statistical informarion of data

Volume     Avg Price  Total Sales Value  Discount Rate (%)  \
count  450.000000    450.000000         450.000000         450.000000   
mean     5.066667  10453.433333       33812.835556          15.155242   
std      4.231602  18079.904840       50535.074173           4.220602   
min      1.000000    290.000000         400.000000           5.007822   
25%      3.000000    465.000000        2700.000000          13.965063   
50%      4.000000   1450.000000        5700.000000          16.577766   
75%      6.000000  10100.000000       53200.000000          18.114718   
max     31.000000  60100.000000      196400.000000          19.992407   

       Discount Amount  Net Sales Value  
count       450.000000       450.000000  
mean       3346.499424     30466.336131  
std        4509.902963     46358.656624  
min          69.177942       326.974801  
25%         460.459304      2202.208645  
50%         988.933733      4677.788059  
75%        5316.495427     47847.912852  
max       25738.022194    179507.479049

In [33]:
data.describe().T[['mean','std','50%']] #Display the mean, standard deviation and median of the data .

mean           std          50%
Volume                 5.066667      4.231602     4.000000
Avg Price          10453.433333  18079.904840  1450.000000
Total Sales Value  33812.835556  50535.074173  5700.000000
Discount Rate (%)     15.155242      4.220602    16.577766
Discount Amount     3346.499424   4509.902963   988.933733
Net Sales Value    30466.336131  46358.656624  4677.788059

In [34]:
df = data.describe().T[['mean','50%','std']]
df.rename(columns={'50%':'Median','mean':'Mean','std':'Standard Deviation'},inplace=True) #rename the columns names
df1 = df
df1

Mean       Median  Standard Deviation
Volume                 5.066667     4.000000            4.231602
Avg Price          10453.433333  1450.000000        18079.904840
Total Sales Value  33812.835556  5700.000000        50535.074173
Discount Rate (%)     15.155242    16.577766            4.220602
Discount Amount     3346.499424   988.933733         4509.902963
Net Sales Value    30466.336131  4677.788059        46358.656624

In [35]:
for i in data.describe().columns:
    print(f"""Mode of {i}:
          {data[i].mode()}\n""") #Find and Display the mode of every colummn

Mode of Volume:
          0    3
Name: Volume, dtype: int64

Mode of Avg Price:
          0     400
1     450
2     500
3    1300
4    8100
Name: Avg Price, dtype: int64

Mode of Total Sales Value:
          0    24300
Name: Total Sales Value, dtype: int64

Mode of Discount Rate (%):
          0       5.007822
1       5.055218
2       5.059801
3       5.072124
4       5.084108
         ...    
445    19.940610
446    19.965340
447    19.976879
448    19.982257
449    19.992407
Name: Discount Rate (%), Length: 450, dtype: float64

Mode of Discount Amount:
          0         69.177942
1         73.025199
2         93.649161
3         94.682735
4        102.705757
           ...     
445    17900.983733
446    21153.498820
447    21496.675367
448    25328.224204
449    25738.022194
Name: Discount Amount, Length: 450, dtype: float64

Mode of Net Sales Value:
          0         326.974801
1         330.822058
2         466.350839
3         485.317265
4         496.607104
           ...   

In [36]:
df.insert(3,"Mode",[3,[400,450,500,1300,8100],24300,17.51,735.71,'None'],True) #inserting a new columns to table
df #display mean , standard deviation , median , mode of data

Mean       Median  Standard Deviation  \
Volume                 5.066667     4.000000            4.231602   
Avg Price          10453.433333  1450.000000        18079.904840   
Total Sales Value  33812.835556  5700.000000        50535.074173   
Discount Rate (%)     15.155242    16.577766            4.220602   
Discount Amount     3346.499424   988.933733         4509.902963   
Net Sales Value    30466.336131  4677.788059        46358.656624   

                                          Mode  
Volume                                       3  
Avg Price          [400, 450, 500, 1300, 8100]  
Total Sales Value                        24300  
Discount Rate (%)                        17.51  
Discount Amount                         735.71  
Net Sales Value                           None

**Volume:** The average volume is low at 5.07 with a small standard deviation, indicating consistent low volume sales across the dataset.

**Avg Price:** There is a significant difference between the mean and median prices, suggesting a skewed distribution with some very high values inflating the average.

**Total Sales Value:** Similar to the average price, the high standard deviation and difference between the mean and median indicate a few very high sales values.

**Discount Rate (%):** The average discount rate is around 15%, with a relatively low standard deviation, suggesting that discount rates are fairly consistent across sales.

**Discount Amount:** The mean discount amount is significantly higher than the median, indicating that while most discounts are moderate, some large discounts skew the average higher.

**Net Sales Value:** Specific statistics are not displayed, but this typically reflects sales after discounts.

# Data Visualization

# Histograms:

In [37]:
#Making plot of data with the column volume, 20 bins and box plot above the histogram
fig = px.histogram(data, x='Volume',nbins=20,template='plotly_dark',text_auto=True)
fig.update_traces(marker_line_color='yellow',marker_line_width=1.2) #Mark edges of histogram with white color
fig.update_layout(title = 'Distribution of volume',  #title of plot
                  title_x = 0.5,                     #position of title
                  xaxis_title='Volume',              #title for x-axis
                  yaxis_title='Frequency')           #title for y-axis
fig.show() #Displaying histogram

**From the above histogram plot, I conclude that:**

*   The column is log-normally distributed and right-skewed.
*   The column contains outliers.
*   It contains extreme values, but there are no unusual distributions within the column.
*   Initially, the first bin shows low values, which then suddenly increase and gradually decrease in the remaining bins.
*   There are no outliers below the lower bound.
*   The outliers are present above the upper bound.


In [38]:
fig = px.histogram(data, x='Avg Price',nbins = 20, text_auto=True,template='plotly_dark')
fig.update_traces(marker_line_color='yellow',marker_line_width=1.2)
fig.update_layout(title='Distribution of Avg Prices',
                  title_x=0.5,
                  xaxis_title='Avg Prices',
                  yaxis_title='Frequency')
fig.show()

**Based on the histogram plot, I conclude the following:**

*  The distribution of the column follows a Poisson distribution.
*  There are outliers at the upper bound.
*  Some bins have no values, which may indicate an unusual distribution.
*  The first bin contains the most values, while the remaining bins have very few.
*  There are no outliers below the lower bound.
* Outliers are present above the upper bound.

















In [39]:
fig = px.histogram(data, x='Total Sales Value',nbins=10,text_auto=True,template='plotly_dark')
fig.update_traces(marker_line_width=1.2,marker_line_color='yellow')
fig.update_layout(title='Distribution of Total Sales Value',
                 title_x=0.5,
                 xaxis_title='Total Sales Value',
                 yaxis_title='Frequency')
fig.show()

**Based on the histogram plot, I conclude that:**

*  This data follows a Poisson distribution.
*  It shares similar properties with the previous column, but there is no unusual distribution in this column.


In [40]:
fig = px.histogram(data, x='Discount Rate (%)',nbins=20, text_auto=True,template='plotly_dark')
fig.update_traces(marker_line_width=1.2,marker_line_color='yellow')
fig.update_layout(title='Distribution Discount Rate (%)',
                 title_x=0.5,
                 xaxis_title='Discount Rate (%)',
                 yaxis_title='Frequency')
fig.show()

**Based on the histogram plot, I conclude that:**

* The column is log-normally distributed and left-skewed.
* There are outliers present.
* There are no unusual distributions in the column.
* Low values are distributed in the initial bins, while higher values are distributed in the remaining bins.
* There are no outliers at the upper bound.
* Outliers are present below the lower bound.







In [41]:
fig = px.histogram(data, x='Discount Amount',nbins=15,template='plotly_dark',text_auto=True)
fig.update_traces(marker_line_color='yellow',marker_line_width=1.2)
fig.update_layout(title='distribution of Discount Amount',
                 title_x=0.5,
                 xaxis_title='Discount Amount',
                 yaxis_title='Frequency')
fig.show()

**Based on the above histogram plot, I conclude that:**

* The column follows a Poisson distribution.
* More values are distributed in the initial bin, then the number of distributed values decreases.
* At the upper bound, a small number of values are distributed, which may be outliers.
* Outliers are present at the upper bound, but there are no outliers at the lower bound.

In [42]:
fig = px.histogram(data, x='Net Sales Value',nbins=20,text_auto=True,template='plotly_dark')
fig.update_traces(marker_line_width=1.2,marker_line_color='yellow')
fig.update_layout(title='Net Sales Value',
                 title_x=0.5,
                 xaxis_title='Net Sales Value',
                 yaxis_title='Frequency')
fig.show()

**Based on the above histogram, I conclude that:**

* The column follows a Poisson distribution.
* There is no unusual distribution in the column.

# Boxplots:

In [43]:
fig = px.box(data, x='Volume',template='plotly_dark')
fig.update_layout(title='Box Plot for Volume',
                 title_x=0.5)
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the upper bound.
* More values are distributed above the median.
* There are a total of 12 outlier values in this column.
* The upper bound is 10, and the lower bound is 1.


In [44]:
fig = px.box(data, x='Avg Price',template='plotly_dark')
fig.update_layout(title='Box plot for Avg Price',
                 title_x=0.5)
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the upper bound.
* The majority of values are distributed above the median.
* There are a total of 4 outliers in this column.
* The upper bound is 20.1k, and the lower bound is 290.





In [45]:
fig = px.box(data, x='Total Sales Value',template='plotly_dark')
fig.update_layout(title='Box Plot for Total Sales Value',
                 title_x=0.5)
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the upper bound.
* The majority of values are distributed above the median.
* There are a total of 11 outliers in this column.
* The upper bound is 120.6k, and the lower bound is 400.

In [46]:
fig = px.box(data, x='Discount Rate (%)',template='plotly_dark')
fig.update_layout(title='Box Plot for Discount Rate (%)',
                 title_x=0.5)
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the lower bound.
* The majority of values are distributed above the median.
* There are several outliers in this column.
* The upper bound is 19.99, and the lower bound is 5.00.

In [47]:
fig = px.box(data, x='Discount Amount',template='plotly_dark')
fig.update_layout(title='Box Plot for Discount Amount',
                 title_x=0.5)
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the upper bound.
* The majority of values are distributed above the median.
* The upper bound is 25.73k, and the lower bound is 69.17.

In [48]:
fig = px.box(data, x='Net Sales Value',template='plotly_dark')
fig.update_layout(title='Net Sales Value',
                 title_x=0.5,
                 xaxis_title='Net Sales Value')
fig.show()

**Based on the above Box Plot, I conclude that:**
* There are outliers at the upper bound.
* The majority of values are distributed above the median.
* The upper bound is 179.50k, and the lower bound is 326.97.

# Bar Chart Analysis for Categorical Column:

In [49]:
Categorical_Columns = list(set(data.columns)-set(data.describe().columns)) #Get the Categorical columns in the data
Categorical_Columns #Displaying the categorical columns

['SKU', 'Brand', 'BU', 'Day', 'City', 'Model', 'Date']

In [50]:
df1 = data['City'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
df1

City  count
0    C    450

In [51]:
fig = px.bar(df1,x='City',y='count',text_auto=True,template='plotly_dark',color='City') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For City', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**
* There is only one unique value in this column.
* The count of this unique value is 450.

In [52]:
df1 = data['Date'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='Date',y='count',text_auto=True,template='plotly_dark',color='Date') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For Date', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**

* All categories in this column are equally distributed.
* The count for each unique value is 30.

In [53]:
df1 = data['Day'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='Day',y='count',text_auto=True,template='plotly_dark',color='Day') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For Day', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**
* There are a total of 7 unique values in this column, representing the days of the week.
* All values are distributed equally, except for Thursday, which has a different count.
* The highest number of sales occur on Thursday.

In [54]:
df1 = data['SKU'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='SKU',y='count',text_auto=True,template='plotly_dark',color='SKU') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For SKU', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**
* All categories in this column are equally distributed.
* The count for each unique value is 15.

In [55]:
df1 = data['Model'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='Model',y='count',text_auto=True,template='plotly_dark',color='Model') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For Model', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**
* All categories in this column are equally distributed.
* The count for each unique value is 15.

In [56]:
df1 = data['BU'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='BU',y='count',text_auto=True,template='plotly_dark',color='BU') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For BU', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**

* All categories in this column are equally distributed.
* The count for each unique value is 150.
* There are a total of 3 categories.

In [57]:
df1 = data['Brand'].value_counts().reset_index() #Getting unique values in the column with respective their count
df1 = pd.DataFrame(df1) #creating dataframe for unique values in the column with respective their count
fig = px.bar(df1,x='Brand',y='count',text_auto=True,template='plotly_dark',color='Brand') #Creating a plot for categorical column with respective their count
fig.update_traces(textposition='outside') #Put text position outside of the bar
fig.update_layout(title='Bar Chart For Brand', #title of the plot
                 title_x=0.5) #title position in the plot
fig.show() #displaying the plot

**Based on the above bar plot, I conclude that:**

* The categories in this column are unequally distributed.
* The brand Jeera has the highest number of sales.

# Standardization of Numerical Variables


* Standardization involves transforming raw data values to a standard scale with a mean of 0 and a standard deviation of 1.
*  This is achieved using the formula:
z=x-mu/sigma.
* Standardization helps reduce the effect of extreme values, improves model performance, and makes data more comparable across different datasets.

* It is commonly used in machine learning, finance, and scientific research to preprocess data before analysis or modeling.

In [58]:
df #Before making z normalization

Mean       Median  Standard Deviation  \
Volume                 5.066667     4.000000            4.231602   
Avg Price          10453.433333  1450.000000        18079.904840   
Total Sales Value  33812.835556  5700.000000        50535.074173   
Discount Rate (%)     15.155242    16.577766            4.220602   
Discount Amount     3346.499424   988.933733         4509.902963   
Net Sales Value    30466.336131  4677.788059        46358.656624   

                                          Mode  
Volume                                       3  
Avg Price          [400, 450, 500, 1300, 8100]  
Total Sales Value                        24300  
Discount Rate (%)                        17.51  
Discount Amount                         735.71  
Net Sales Value                           None

In [59]:
df.describe().columns #Get numerical columns

Index(['Mean', 'Median', 'Standard Deviation'], dtype='object')

In [60]:
for x in df.describe().columns:
    l = [] #Initilizing the list to append all values to this list later update values using this list
    for y in df[x]:
        mean = df[x].mean() #Calculating the mean of the column
        std = df[x].std() #Calculating the standard deviation of the column
        z = (y-mean)/std #calculating the z value
        l.append(round(z,2)) #append calculated z values to list
    df[x].update(l) #Updating the column with list
df #Displaying the column after normalization

Mean       Median  Standard Deviation  \
Volume                 5.066667     4.000000            4.231602   
Avg Price          10453.433333  1450.000000        18079.904840   
Total Sales Value  33812.835556  5700.000000        50535.074173   
Discount Rate (%)     15.155242    16.577766            4.220602   
Discount Amount     3346.499424   988.933733         4509.902963   
Net Sales Value    30466.336131  4677.788059        46358.656624   

                                          Mode  
Volume                                       3  
Avg Price          [400, 450, 500, 1300, 8100]  
Total Sales Value                        24300  
Discount Rate (%)                        17.51  
Discount Amount                         735.71  
Net Sales Value                           None

# Conversion of Categorical Data into Dummy Variables

* One-hot encoding involves converting categorical data into dummy variables.
* This transformation is necessary because algorithms such as linear regression can only handle continuous or numerical data.
* Categorical data cannot be directly processed by these models.
* One-hot encoding transforms categorical data into a binary format, where each category is represented as a separate binary column.
* This allows algorithms to differentiate between distinct categories and enables accurate modeling and prediction of categorical relationships, thereby improving the overall performance of machine learning models.  



In [61]:
d = pd.get_dummies(data,columns=Categorical_Columns,dtype=int)
d.head()

Volume  Avg Price  Total Sales Value  Discount Rate (%)  Discount Amount  \
0      15      12100             181500          11.654820     21153.498820   
1      10      10100             101000          11.560498     11676.102961   
2       7      16100             112700           9.456886     10657.910157   
3       6      20100             120600           6.935385      8364.074702   
4       3       8100              24300          17.995663      4372.946230   

   Net Sales Value  SKU_F01  SKU_F02  SKU_F03  SKU_F04  ...  Date_06-04-2021  \
0    160346.501180        0        0        0        0  ...                0   
1     89323.897039        0        0        0        0  ...                0   
2    102042.089843        0        0        0        0  ...                0   
3    112235.925298        0        0        0        0  ...                0   
4     19927.053770        0        0        0        0  ...                0   

   Date_07-04-2021  Date_08-04-2021  Date_09-04-2021  Date_10-04-2021  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Date_11-04-2021  Date_12-04-2021  Date_13-04-2021  Date_14-04-2021  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Date_15-04-2021  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 101 columns]

# Conclusion

* Data preprocessing is a crucial step in data analysis and machine learning, as it enables accurate and reliable modeling.
* Standardization and one-hot encoding are two essential preprocessing techniques.
* Standardization scales data to a common range to prevent features with large ranges from dominating others.
* One-hot encoding converts categorical variables into binary vectors, allowing them to be processed by models.
* These techniques help prevent data leakage, improve model interpretability, and enable more efficient computation.
* By applying standardization and one-hot encoding, you can ensure that your data is clean, consistent, and ready for analysis and modeling.  



1. **Sales Volume Distribution**

    * **Mean Volume:** 5.07 units
    * **Standard Deviation:** 4.23 units
    * **Range:** 1 to 450 units.

The histogram shows a high frequency at the lower end, indicating that most sales volumes are low.
2. **Average Price Distribution**

* **Mean Average Price:** 10,453.40
* **Standard Deviation:** 18,079.90
* **Range:** 290 to 60,100

The histogram reveals a right-skewed distribution, suggesting that while most products are priced lower, a few high-priced items significantly raise the average price.

3. **Total Sales Value Distribution**

* **Mean Total Sales Value:** 33,812.80
* **Standard Deviation:** 50,535.10
* **Range:** 400 to 196,400

The histogram indicates a right-skewed distribution, similar to average prices, with most sales values clustering at the lower end.

4. **Discount Rate Distribution**

* **Mean Discount Rate:** 15.16%
* **Standard Deviation:** 4.22%
* **Range:** 5.01% to 450%  

The histogram shows a concentration of discount rates at lower percentages, with a few outliers showing extremely high discount rates.

5. **Net Sales Value Distribution**

* **Mean Net Sales Value:** 30,466.34
* **Standard Deviation:** 46,358.66
* **Range:** 326.97 to 179,507.48

The histogram for net sales values shows a distribution with multiple peaks, indicating variability in net sales across different values.



**Data Preprocessing:**
Data preprocessing is a crucial step in data analysis and machine learning, as it ensures accurate and reliable modeling. Two essential preprocessing techniques are:

* **Standardization:** This technique scales data to a common range to prevent features with large ranges from dominating others. It transforms data to have a mean of 0 and a standard deviation of 1.

* **One-Hot Encoding:** This technique converts categorical variables into binary vectors, allowing them to be processed by models.

These preprocessing techniques help prevent data leakage, improve model interpretability, and enable more efficient computation. By applying standardization and one-hot encoding, you ensure that your data is clean, consistent, and ready for analysis and modeling.